internal/validation/mock_trace_validation.go

```diff
@@ -56,57 +57,94 @@ var (
 	agentRegex     = regexp.MustCompile(`^opentelemetry-[a-zA-Z]+ [0-9]+\.[0-9]+\.[0-9]+; google-cloud-trace-exporter [0-9]+\.[0-9]+\.[0-9]+$`)
 )
 
+// SpanData wraps all the span data on the server into a struct.
+type SpanData struct {
+	// If a batch has a bad span, we don't write batch to memory, but still want
+	// info on them for summary, so need SpansSummary
+	SpansSummary      []*cloudtrace.Span
+	UploadedSpanNames map[string]struct{}
+	UploadedSpans     []*cloudtrace.Span
+	Mutex             sync.Mutex
+}
+
 // ValidateSpans checks that the spans conform to the API requirements.
 // That is, required fields are present, and optional fields are of the correct form.
-func ValidateSpans(requestName string, spans ...*cloudtrace.Span) error {
+// If any violations are detected, the errors will be added to the result table.
+func ValidateSpans(requestName string, spanData *SpanData, spans ...*cloudtrace.Span) error {
```

Comment:
Yeah I agree, the problem is `spanData` is holds the data for as long as the server runs (in which `ValidateSpans` can be called many times), while the return value `error` just returns the error for the current call of `BatchWriteSpans` or `CreateSpan`. I'm not sure if there is an easy way to extract the current error from `spanData`, so I figured this might be the easiest solution for now